cmd/snap-bootstrap/partition/lsblk.go

```diff
@@ -0,0 +1,132 @@
+ *
+ */
+
+package partition
+
+import (
+	"encoding/json"
+	"fmt"
+	"os/exec"
+	"strconv"
+
+	"github.com/snapcore/snapd/osutil"
+)
+
+var (
+	procSelfMountInfo = osutil.ProcSelfMountInfo
+)
+
+// lsblkFilesystemInfo represents the lsblk --fs JSON output format.
+type lsblkFilesystemInfo struct {
+	BlockDevices []lsblkBlockDevice `json:"blockdevices"`
+}
+
+type lsblkBlockDevice struct {
+	Name          string             `json:"name"`
+	FSType        string             `json:"fstype"`
+	Label         string             `json:"label"`
+	UUID          string             `json:"uuid"`
+	Mountpoint    string             `json:"mountpoint"`
+	PartitionUUID string             `json:"partuuid"`
+	Children      []lsblkBlockDevice `json:"children"`
+}
+
+func lsblckFsInfo(opts ...string) (*lsblkFilesystemInfo, error) {
+	args := append(
+		[]string{
+			"--json",
+			// same options as --fs, but also with partuuid
+			"-o", "NAME,FSTYPE,LABEL,UUID,MOUNTPOINT,PARTUUID",
+		},
+		opts...,
+	)
+	output, err := exec.Command("lsblk", args...).CombinedOutput()
+	if err != nil {
+		return nil, osutil.OutputErr(output, err)
+	}
+
+	var info lsblkFilesystemInfo
+	if err := json.Unmarshal(output, &info); err != nil {
+		return nil, fmt.Errorf("cannot parse lsblk output: %v", err)
+	}
+
+	return &info, nil
+}
+
+func filesystemInfo(node string) (*lsblkFilesystemInfo, error) {
+	return lsblckFsInfo(node)
+}
+
+func filesystemDeviceNumberInfo(majorNum string) (*lsblkFilesystemInfo, error) {
+	return lsblckFsInfo("--include", majorNum)
+}
+
+// Disk is a single physical disk device that contains partitions.
+type Disk struct {
+	dev lsblkBlockDevice
+}
+
+// DiskFromMountPoint finds a matching Disk for the specified mount point.
+func DiskFromMountPoint(mountpoint string) (*Disk, error) {
+	// first get the mount entry for the mountpoint
+	mounts, err := osutil.LoadMountInfo(procSelfMountInfo)
+	if err != nil {
+		return nil, err
+	}
+	found := false
+	var devMajor int
+	for _, mount := range mounts {
+		if mount.MountDir == mountpoint {
+			devMajor = mount.DevMajor
+			found = true
+			break
+		}
+	}
+	if !found {
+		return nil, fmt.Errorf("couldn't find mountpoint %q", mountpoint)
+	}
+
+	info, err := filesystemDeviceNumberInfo(strconv.Itoa(devMajor))
+	if err != nil {
+		return nil, err
+	}
+
+	switch {
+	case len(info.BlockDevices) == 0:
+		// unknown device number to lsblk
+		return nil, fmt.Errorf("lsblk couldn't find device with major number %d", devMajor)
+	case len(info.BlockDevices) > 1:
+		// unclear how this could happen? one mount point from multiple devices?
+		return nil, fmt.Errorf("internal error: multiple block devices for single mountpoint")
+	}
+
+	return &Disk{dev: info.BlockDevices[0]}, nil
+}
+
+// FindMatchingParitionUUID finds the partition uuid for a partition matching
+// the specified label on the disk.
+func (disk *Disk) FindMatchingParitionUUID(label string) (string, error) {
```

Comment:
Typo here, Parition vs Partition.

Reply:
Actually I did this many times in the installer code. It seems to be a very frequent typo.